File: CSharp/LeetCode/Solutions/0704-binary-search.cs
using LeetCode.Solutions.Shared;

namespace LeetCode.Solutions;

internal class Solution0704: IRunProgram
{
    public void Run()
    {
        Search([-1, 0, 3, 5, 9, 12], 9).Print();
    }
    public int Search(int[] nums, int target)
    {
        int left = 0, right = nums.Length - 1;
        while (left <= right)
        {
            int mid = (left + right) >> 1;
            if (nums[mid] < target)
            {
                left = mid + 1;
            }
            else if (nums[mid] > target)
            {
                right = mid - 1;
            }
            else
            {
                return mid;
            }
        }
        return -1;
    }
}

File: Python/0704-binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r :
            mid = (l + r) // 2
            if nums[mid] < target :
                l = mid + 1
            elif nums[mid] > target :
                r = mid - 1
            else :
                return mid

        return -1
        
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.rec(nums, target, 0, (len(nums) - 1) // 2, len(nums) - 1)
    def rec(self, nums: List[int], target: int, l : int, mid : int, r : int) -> int:
        if l >= r :
            return l if nums[l] == target else -1
        
        if nums[mid] < target :
            return self.rec(nums, target, mid + 1, (mid + 1 + r) // 2, r)
        else :
            return self.rec(nums, target, l, (mid + l) // 2, mid)
